README.md
# py3server

py3server is a web application framework, based on decorators patterns.

# Documentation

You can find package API documentation [here](docs/README.md).

# Installing

Install and update using pip:

```shell
    $ pip install -U py3server
```

# Example

```python
# main.py
from py3server.booter import Booter


def main():
    """ """
    Booter(__file__).run()


if __name__ == '__main__':
    main()
```

```python
# user_controller.py
from py3server.web.controllers.rest_controller import RestController
from py3server.web.mapping.post_mapping import PostMapping
from py3server.web.params.body import Body
from py3server.web.response import Response
from .user_service import UserService


@RestController('/users')
class UserController():
    user_service: UserService

    @PostMapping()
    def create_user(self, body: Body):
        user = self.user_service.add_user(body)
        return Response(200, user)
```

```python
# user_service.py
from py3server.app.service import Service
from .user_repository import UserRepo


@Service()
class UserService():
    user_repo: UserRepo

    def add_user(self, data):
        firstname = data["firstname"]
        lastname = data["lastname"]
        return self.user_repo.add_user(firstname, lastname)
```

```python
# user_repo.py
from py3server.app.repository import Repository


@Repository()
class UserRepo():
    users = []

    def add_user(self, firstname, lastname):
        user = {
            "firstname": firstname,
            "lastname": lastname
        }
        self.users.append(user)
        return user
```

CHANGELOG.md
# CHANGELOG

[v1.0.0](#v1-0-0)

## <span name="v1-0-0">1.0.0</span>

#### Added

- Start http server with [Booter](docs/README.md#booter)
- Register components:
    - [@RestController](docs/README.md#rest-controller)
    - [@Service](docs/README.md#service)
    - [@Repository](docs/README.md#repository)
    - [@ExceptionHandler](docs/README.md#exception-handler)
- HTTP methods mapping:
    - [@GetMapping](docs/README.md#get-mapping)
    - [@PostMapping](docs/README.md#post-mapping)
    - [@PutMapping](docs/README.md#put-mapping)
    - [@DeleteMapping](docs/README.md#delete-mapping)
- Params:
    - [Body](docs/README.md#body)
    - [@PathParam](docs/README.md#path-param)
    - [@IntPP](docs/README.md#int-path-param)
    - [@QueryParam](docs/README.md#query-param)
    - [@IntQP](docs/README.md#int-query-param)
- Built response object:
    - [Response](docs/README.md#base-response)

#### Changed

#### Deprecated

#### Removed

#### Fixed

#### Security

tests/conftest.py
import pytest

from py3server.context import Context
from py3server.utils.singleton import SingletonMeta


@pytest.fixture(autouse=True)
def clean_up_context():
    """
    Before each tests delete all saved instances.
    :return:
    """
    Context._instance = None


tests/test_booter.py
import importlib
import os
import tempfile
import typing as t
from unittest.mock import call

from pytest_mock import MockerFixture

from py3server.booter import Booter


def test_booter_init():
    # GIVEN
    root_file: str = "my/root/folder/file.py"

    # WHEN
    booter: Booter = Booter(root_file)

    # THEN
    assert booter.root_file == root_file
    assert booter.address == '0.0.0.0'
    assert booter.port == 8080


def test_booter_discover(mocker: MockerFixture) -> None:
    # GIVEN
    mocker.patch('importlib.import_module')

    temp_dir: tempfile.TemporaryDirectory = tempfile.TemporaryDirectory()
    _, temp_dir_name = os.path.split(temp_dir.name)
    main_file: str = 'main.py'

    valid_files: t.List[str] = ['file1.py', 'file2.py', 'file3.py']
    invalid_files: t.List[str] = ['invalid_file2.py.txt', 'invalid_file3']

    os.mkdir(os.path.join(temp_dir.name, 'not_a_file'))

    for file in [main_file] + valid_files + invalid_files:
        path: str = os.path.join(temp_dir.name, file)
        with open(path, "w+") as fp:
            pass

    booter: Booter = Booter(os.path.join(temp_dir.name, main_file))

    # WHEN
    booter.discover()

    # THEN
    importlib.import_module.assert_has_calls([
        call(f'{temp_dir_name}.{file.replace(".py", "")}') for file in valid_files
    ], any_order=True)



e2e/test_users_api.py
import pytest
import requests


@pytest.mark.run(order=1)
def test_users_api_create_user():
    # GIVEN
    url: str = 'http://localhost:8080/users'
    user_data: object = {
        'firstname': 'user1_firstname',
        'lastname': 'user1_lastname',
    }

    # WHEN
    res: requests.Response = requests.post(url, json=user_data)

    # THEN
    assert res.status_code == 200
    assert res.json() == {
        'firstname': 'user1_firstname',
        'lastname': 'user1_lastname',
        'age': 0
    }


@pytest.mark.run(order=2)
def test_users_api_create_user_with_age():
    # GIVEN
    url: str = 'http://localhost:8080/users'
    user_data: object = {
        'firstname': 'user2_firstname',
        'lastname': 'user2_lastname',
        'age': 25
    }

    # WHEN
    res: requests.Response = requests.post(url, json=user_data)

    # THEN
    assert res.status_code == 200
    assert res.json() == {
        'firstname': 'user2_firstname',
        'lastname': 'user2_lastname',
        'age': 25
    }


@pytest.mark.run(order=3)
def test_users_api_get_all_users():
    # GIVEN
    url: str = 'http://localhost:8080/users'

    # WHEN
    res: requests.Response = requests.get(url)

    # THEN
    assert res.status_code == 200
    assert res.json() == [
        {
            'firstname': 'user1_firstname',
            'lastname': 'user1_lastname',
            'age': 0
        },
        {
            'firstname': 'user2_firstname',
            'lastname': 'user2_lastname',
            'age': 25
        }
    ]


@pytest.mark.run(order=4)
def test_users_api_get_user_2():
    # GIVEN
    url: str = 'http://localhost:8080/users/2'

    # WHEN
    res: requests.Response = requests.get(url)

    # THEN
    assert res.status_code == 200
    assert res.json() == {
        'firstname': 'user2_firstname',
        'lastname': 'user2_lastname',
        'age': 25
    }


@pytest.mark.run(order=5)
def test_users_api_update_user_2():
    # GIVEN
    url: str = 'http://localhost:8080/users/2'
    data: object = {
        'firstname': 'user2_updated_firstname'
    }

    # WHEN
    res: requests.Response = requests.put(url, json=data)

    # THEN
    assert res.status_code == 200
    assert res.json() == {
        'firstname': 'user2_updated_firstname',
        'lastname': 'user2_lastname',
        'age': 25
    }


@pytest.mark.run(order=6)
def test_users_api_delete_user_2():
    # GIVEN
    url: str = 'http://localhost:8080/users/2'

    # WHEN
    res: requests.Response = requests.delete(url)

    # THEN
    assert res.status_code == 204


e2e/project/controllers/user_controller.py
from project.service.user_service import UserService
from py3server.web.controllers.rest_controller import RestController
from py3server.web.mapping.delete_mapping import DeleteMapping
from py3server.web.mapping.get_mapping import GetMapping
from py3server.web.mapping.post_mapping import PostMapping
from py3server.web.mapping.put_mapping import PutMapping
from py3server.web.params.body import Body
from py3server.web.params.int_path_param import IntPP
from py3server.web.response import Response


@RestController('/users')
class UserController(object):
    user_service: UserService

    @PostMapping()
    def create_user(self, body: Body):
        """
        User creation endpoint;
        required:
            - firstname
            - lastname
        optional:
            - age
          """
        errors = []
        required_fields = ['firstname', 'lastname']
        for required_field in required_fields:
            if not body.get(required_field):
                errors.append({
                    'field': required_field,
                    'code': 'REQUIRED'
                })

        if len(errors) > 0:
            return Response(400, errors)

        firstname, lastname, age = [
            body['firstname'],
            body['lastname'],
            body.get('age', 0)
        ]

        user = self.user_service.add_user(firstname, lastname, age)

        return Response(200, user)

    @GetMapping()
    def get_users(self):
        users = self.user_service.get_users()
        return Response(200, users)

    @GetMapping('/{id}')
    def get_user(self, user_id: IntPP('id')):
        user = self.user_service.get_user(user_id)
        return Response(200, user)

    @PutMapping('/{id}')
    def update_user(self, user_id: IntPP('id'), body: Body):
        user = self.user_service.update_user(user_id, body)
        return Response(200, user)

    @DeleteMapping('/{id}')
    def delete_user(self, user_id: IntPP('id')):
        self.user_service.remove_user(user_id)
        return Response(204)
